packages/augur-state/src/db/DB.ts

```diff
@@ -109,32 +110,60 @@ export class DB<TBigNumber> {
    */
   public async sync(augur: Augur<TBigNumber>, chunkSize: number, blockstreamDelay: number): Promise<void> {
     // Sync generic event types
-    let dbSyncPromises = [];
+    const maxConcurrency = 20;
+    let q = queue(function(task: any, callback: any) {
+      task.dbController.syncableDatabases[task.dbIndex].sync(task.augur, task.chunkSize, task.blockstreamDelay, task.dbController.syncStatus.defaultStartSyncBlockNumber);
+      callback();
+    }, maxConcurrency);
     for (let dbIndex in this.syncableDatabases) {
-      dbSyncPromises.push(this.syncableDatabases[dbIndex].sync(augur, chunkSize, blockstreamDelay, this.syncStatus.defaultStartSyncBlockNumber));
+      q.push(
+        {
+          dbController: this, 
+          dbIndex,
+          augur, 
+          chunkSize,
+          blockstreamDelay,
+        },
+        function(err: Error) {
+          if (err) {
+            console.log(err);
+          } else {
+            console.log("Finished syncing SyncableDB " + dbIndex);
+          }
+        }
+      );
     }
-    // TODO Figure out a way to handle concurrent request limit of 40
-    await Promise.all(dbSyncPromises)
-    .catch(error => {
-      throw error;
-    });
 
     // Sync user-specific event types
     // TODO TokensTransferred should comprise all balance changes with additional metadata and with an index on the to party.
     // Also update topics/indexes for user-specific events once these changes are made to the contracts.
+    q = queue(function(task: any, callback: any) {
+      task.dbController.syncableDatabases[task.dbName].sync(task.augur, task.chunkSize, task.blockstreamDelay, task.dbController.syncStatus.defaultStartSyncBlockNumber);
+      callback();
+    }, maxConcurrency);
     for (let trackedUser of await this.trackedUsers.getUsers()) {
       for (let userSpecificEvent of this.userSpecificEvents) {
         let dbName = this.getDatabaseName(userSpecificEvent.name, trackedUser);
-        dbSyncPromises.push(this.syncableDatabases[dbName].sync(augur, chunkSize, blockstreamDelay, this.syncStatus.defaultStartSyncBlockNumber));
+        q.push(
+          {
+            dbController: this, 
+            dbName,
+            augur, 
+            chunkSize,
+            blockstreamDelay,
+          },
+          function(err: Error) {
+            if (err) {
+              console.log(err);
+            } else {
+              console.log("Finished syncing SyncableDB " + dbName);
+            }
+          }
+        );
       }
     }
-    // TODO Figure out a way to handle concurrent request limit of 40
-    await Promise.all(dbSyncPromises)
-    .catch(error => {
-      throw error;
```

Comment:
Is there a way to make the `sync` function `await` until the queues empty? My tests rely on `sync` passing along database exceptions. I don't like the idea of logging without being able to respond to the exceptions higher up anyway.

Reply:
Sure, that sounds good. I'll make that change and add it to this PR on Monday.